[Documentation] Trying to rewrite typeclasses-new (#8147)
lampepfl/dotty: docs/docs/reference/contextual/typeclasses-new.md

```scala
given as Monoid[String] {
  def (x: String) combine (y: String): String = x.concat(y)
  def unit: String = ""
}
```

Whereas for the type `Int` one could write the following:
```scala
given as Monoid[Int] {
  def (x: Int) combine (y: Int): Int = x + y
  def unit: Int = 0
}
```

This monoid can now be used as _context bound_ in the following `combineAll` method:

```scala
def combineAll[T: Monoid](xs: List[T]): T =
    xs.foldLeft(summon[Monoid[T]].unit)(_ combine _)
```

To get rid of the `summon[...]` we can define a `Monoid` object as follows:

```scala
object Monoid {
  def apply[T] with (m: Monoid[T]) = m
}
```

Which would allow to re-write the `combineAll` method this way: 

```scala
def combineAll[T: Monoid](xs: List[T]): T =
    xs.foldLeft(Monoid[T].unit)(_ combine _)
```

We can also benefit from [extension methods](extension-methods-new.html) to make this `combineAll` function accessible as a method on the `List` type:

```scala
def [T: Monoid](xs: List[T]).combineAll: T =
  xs.foldLeft(Monoid[T].unit)(_ combine _)  
```

Which allows one to write: 

```scala
assert("ab" == List("a", "b").combineAll)
```
or:
```scala
assert(3 == List(1, 2).combineAll)
```

> 

Reviewer: Thanks for doing this, it's a lot more with the style of the rest of the documentation. If you apply the same style to the Functors and Monads section this will be really nice.

Author: OK I'm not sure I have enough knowledge to re-write the Functor/Monad part in the same fashion (you need to understand properly to explain properly) but I will definitely give it a try.
If I'm struggling I'll ping you, so that you know you can merge this as is (best if the ennemy of good ;) ). But that's worth trying.

```scala
given as Monoid[Int] {
  def (x: Int) combine (y: Int): Int = x + y
  def unit: Int = 0
}
```

This monoid can now be used as _context bound_ in the following `combineAll` method:

```scala
def combineAll[T: Monoid](xs: List[T]): T =
    xs.foldLeft(summon[Monoid[T]].unit)(_ combine _)
```

To get rid of the `summon[...]` we can define a `Monoid` object as follows:

```scala
object Monoid {
  def apply[T] with (m: Monoid[T]) = m
}
```

Which would allow to re-write the `combineAll` method this way: 

```scala
def combineAll[T: Monoid](xs: List[T]): T =
    xs.foldLeft(Monoid[T].unit)(_ combine _)
```

We can also benefit from [extension methods](extension-methods-new.html) to make this `combineAll` function accessible as a method on the `List` type:

```scala
def [T: Monoid](xs: List[T]).combineAll: T =
  xs.foldLeft(Monoid[T].unit)(_ combine _)  
```

Which allows one to write: 

```scala
assert("ab" == List("a", "b").combineAll)
```
or:
```scala
assert(3 == List(1, 2).combineAll)
```

### Functors:

A `Functor` represents the ability for a type containing zero or more elements to be "mapped over", i.e. apply a function to every of its elements.

> Let's name our "type containing zero or more elements" `F`. It's a [higher-kinded type](http://guillaume.martres.me/publications/dotty-hk.pdf), since it contains elements of an other type.

Reviewer: The higher kinded type implementation described here is out of date, so I suggest that only a high level description should be linked, (which I can't find and official documentation for), it is easier to describe `F` as a Type Constructor, where the type of its values becomes concrete when provided a type argument.

Author: ✅

```scala
given listMonad as Monad[List] {
  def pure[A](x: A): List[A] =
    List(x)
  def [A, B](xs: List[A]).flatMap(f: A => List[B]): List[B] =
    xs.flatMap(f) // let's rely on the existing `flatMap` method of `List`
}
```

`map` implementation is no longer needed.

#### Option

`Option` is an other type having the same kind of behaviour:
* the `map` ability turning `Option[A]` into `Option[B]` if passed a function `f: A => B`
* the `flatMap` ability turning `Option[A]` into `Option[B]` if passed a function `f: A => Option[B]`
* the `pure` ability turning `A` into `Option[A]`

```scala
given optionMonad as Monad[Option] {
  def pure[A](x: A): Option[A] =
    Option(x)
  def [A, B](xs: Option[A]).flatMap(f: A => Option[B]): Option[B] =
    xs.flatMap(f) // let's rely on the existing `flatMap` method of `Option`
}
```

#### The Reader Monad

Another example of a `Monad` is the Reader Monad. It no longer acts on a type like `List` or `Option`, but on a function.
It can be used for example for combining functions that all need the same type of parameter. For instance multiple functions needing access to some configuration, context, environment variables, etc.

> Let us have a `Config` type, and two functions using it:

Reviewer: ```suggestion
To explore the use case for reader, let us define a concrete `Config` trait, and two operations that require a `Config` to be provided:
```

```scala
trait Config
def compute(i: Int)(config: Config): String = ???
def show(str: String)(config: Config): Unit = ???
```

> We may want to combine `compute` and `show` into a single function, accepting a `Config` as parameter, and showing the result of the computation.

Reviewer: ```suggestion
We may want to combine `compute` and `show` into a single operation, accepting a `Config` as parameter, and showing the result of the computation.
```

If we had a `flatMap` function as in the examples above, we would be able to write the following:

```scala
def computeAndShow(i: Int): Config => Unit = compute(i).flatMap(show)
```

> Let's define this `Monad` then. First, we are going to define a type named `ConfigDependent` representing a function that when passed a `Config` produces a `Result`.

Reviewer: ```suggestion
Let's define a `Monad` to chain our operations with `flatMap`. First, we are going to define a type named `ConfigDependent` representing a function that when passed a `Config` produces a `Result`.
```

```scala
trait Config // the Config defined above
type ConfigDependent[Result] = Config => Result
```

The monad will look like this:

```scala
given configDependentMonad as Monad[ConfigDependent]
  def [A, B](r: ConfigDependent[A]).flatMap(f: A => ConfigDependent[B]): ConfigDependent[B] =
    config => f(r(config))(config)
  def pure[A](x: A): ConfigDependent[A] =
    config => x
```

The type `ConfigDependent` can be written using [type lambdas](../new-types/type-lambdas.html):

```scala
type ConfigDependent = [Result] =>> Config => Result
```

Using this syntax would turn the previous `configReaderMonad` into:

```scala
given configDependentMonad as Monad[[Result] =>> Config => Result]
  def [A, B](r: Config => A).flatMap(f: A => Config => B): Config => B =
    config => f(r(config))(config)
  def pure[A](x: A): Config => A =
    config => x
```

> The Reader monad allows to abstract over the `Config` type, named `Ctx` in the following examples. It is therefore _parameterized_ by `Ctx`:

Reviewer: ```suggestion
It is likely that we would like to use this pattern with other kinds of environments than our `Config` trait. The Reader monad allows us to abstract away `Config` as a type _parameter_, named `Ctx` in the following definition:
```

Author: So much better than what I wrote originally, thanks a lot.

```scala
given readerMonad[Ctx] as Monad[[X] =>> Ctx => X] {
  def [A, B](r: Ctx => A).flatMap(f: A => Ctx => B): Ctx => B =
    ctx => f(r(ctx))(ctx)
  def pure[A](x: A): Ctx => A =
    ctx => x
}
```

### Summary

> The definition of a _typeclass_ is expressed in Scala 3 via a `trait`.

Reviewer: ```suggestion
The definition of a _typeclass_ is expressed via a parameterised type with abstract members, such as a `trait`.
```

> The main difference with other traits resides in how these traits are implemented. 

Reviewer: ```suggestion
The main difference between object oriented polymorphism, and ad-hoc polymorphism with _typeclasses_, is how the definition of the _typeclass_ is implemented, in relation to the type it acts upon.
```

> In the case of a _typeclass_ the trait's implementations are expressed through `given ... as` type definitions, and not through classes that `extends` the trait linearly.

Reviewer: ```suggestion
In the case of a _typeclass_, its implementation for a concrete type is expressed through a `given ... as` term definition, which is supplied as an implicit argument alongside the value it acts upon. With object oriented polymorphism, the implementation is mixed into the parents of a class, and only a single term is required to perform a polymorphic operation.
```

> In addition to these given instances, other constructs like extension methods, context bounds and type lambdas allow a concise and natural expression of _typeclasses_.

Reviewer: ```suggestion
To conclude, in addition to given instances, other constructs like extension methods, context bounds and type lambdas allow a concise and natural expression of _typeclasses_.
```

    ctx => x
}
```

> The conjunction of given instances, extension methods and context bounds allow a concise and natural expression of _typeclasses_

Reviewer 2: Stray text?

Reviewer 2: Otherwise LGTM

Author: It was meant to be a sort of conclusion, re-using the previous sentence from the docs (since I think it's accurate). 

I'll try to re-write properly
